### ISSUE TITLE
Improve handling of summary view with semidefinite and indefinite ranges

### ISSUE THREAD

**Issue by pkienzle** [bumps/bumps#85]
For parameters with infinite range the linear scroll bar on the summary view is unusable.

This comes up when you toggle a fit parameter in the GUI that was not set in the original model since it has no fit range set.

Some ideas:

1. fix indefinite ranges to e.g., ±50% of original value
2. use logarithmic bar spanning e.g., 6 orders of magnitude of the original value
3. use nonlinear bar which allows fine tuning (e.g., 1% per step) near the original value, and exponential steps far away
4. use nonlinear bar which allows fine tuning (e.g., 1% per step) if you move slowly but warps exponentially by moving fast
5. as above, but with bar recentered to the new value each time you release slider

Option 1 is easiest: check for indefinite ranges when toggling fittable parameters and make them definite.

Option 4+5 works well for physical devices. For example, a trackpad moves the cursor slowly if you drag slowly but warps if you drag quickly. Option 3+5 may be almost as good, using distance from center instead of speed.

**Comment by pkienzle** [comment]
Note that there is no convenient method for setting a fit range on a non-fittable parameter in the python model. I don't think that will be an issue if we can handle indefinite ranges well.

We may want to restrict the range to e.g., ±50% when fitting even if we have a scheme to easily explore orders of magnitude with the slider. Some optimizers perform much better when they have limits.

**Comment by bmaranville** [comment]
I haven't seen any parameters that need orders of magnitude type exploration in reflectometry, though of course bumps gets used for much more than that.

**Comment by bmaranville** [comment]
I suppose we could try to base it on what the fitter does - if there is a step size that gets assigned to unbounded fit parameters, that would be a logical step size for one tick of the slider, for instance.

**Comment by pkienzle** [comment]
Parameter ranges are defaulting either to unbounded or [0,∞). If the user doesn't set the parameter range in the model it will default to this.

We are already using what the fitter is using:
https://github.com/bumps/bumps/blob/55295a3e2ecb4332e19952b9aef1fdab1db68c3f/bumps/gui/summary_view.py#L218

In the unbounded case this resolves to:
https://github.com/bumps/bumps/blob/55295a3e2ecb4332e19952b9aef1fdab1db68c3f/bumps/bounds.py#L769-L786

**Comment by bmaranville** [comment]
> Note that there is no convenient method for setting a fit range on a non-fittable parameter in the python model. I don't think that will be an issue if we can handle indefinite ranges well.

What do you mean by this?  I know that using e.g. Parameter.range does two things, toggling "fit" and setting the user limits - but when would you want to set a fit range on a non-fittable parameter?

**Comment by pkienzle** [comment]
The `FitProblem` interface does not have a convenient way to set the bounds to use on a fixed parameter in the user decides to toggle it to fitted in the GUI. Any time you set the bounds it will set the parameter to fitted.

To see this, set up a model with some parameter fixed then load this model into the GUI. Clicking `Fit?` on the fixed parameter (i.e., setting the checkbox) will turn it into an unbounded parameter since it doesn't have a default fit range.

A solution is to handle unbounded ranges better and/or we should default the parameter to reasonable (and editable!) bounds when no bounds were set before toggling. I don't think we need to make it easier to define bounds on a fixed parameter in the model.

**Comment by bmaranville** [comment]
I think +/- 1 are reasonable bounds for a parameter - I think this will have less failure modes and edge cases (what if the parameter is near zero, or zero?)  If it were trivially easy, I might even suggest popping up a dialog to encourage users to enter bounds if they toggle a previously fixed parameter to 'fit'.  We could temporarily highlight the automatically-generated bounds if we wanted to make sure the user noticed that they had been applied.

The other thing is that you and I spent a lot of time reworking the Parameter interface in the dataclass_overlay branch, and the limits (user and model) are defined separately from the distribution definitions.  I just glanced at that code again, and in it when you do e.g. Parameter.pm it simply adjusts the (user) bounds and doesn't do anything else, which I believe accomplishes "making it easier to define bounds on a fixed parameter"

**Comment by pkienzle** [comment]
Yes, my point was that setting the range also set `fixed=False`.

A long time ago I considered adding a `fixed=False` option to the range setters. Or maybe method chaining `P.pm(10).fixed`. This was to address a related quirk: fixed parameters can be dragged arbitrarily in the profile view but bounded parameters are constrained to the parameter range. I opted to do nothing. My reasoning was that the user didn't think they needed to fit the parameter when they created the model, so they wouldn't have bothered to set the initial bounds.

±1 would be okay for SLD but not reasonable for length=250, background=1e-6 or fraction=0.2 (though for a fraction the parameter limits of [0,1] would kick in). Maybe add a `default_bounds` method to `Parameter` so we can control it the same way we control parameter limits.

This feels over-engineered. A GUI builder can put appropriate bounds on the fixed parameters according to the type. I'm inclined to ignore the problem, unless you can think of something clever for unbounded ranges.

**Comment by pkienzle** [comment]
log sliders work well for backgrounds and for length scales in scattering. The slider for scale in SANS would also benefit from logarithmic spacing. We could make this a flag for the parameter in the model much like we do for limits.

We still need better behaviour for indefinite limits. Currently the slider is disabled.
